### ISSUE TITLE
Metadata on gdscript resource does not get saved

### ISSUE THREAD

**Issue by donn-xx** [godotengine/godot#84653]
### Godot version

v4.2.beta4.official [93cdacbb0]

### System information

Godot v4.2.beta4 - Ubuntu 22.04.3 LTS 22.04 - X11 - Vulkan (Forward+) - dedicated NVIDIA GeForce RTX 2060 (nvidia; 535.129.03) - Intel(R) Core(TM) i5-9400F CPU @ 2.90GHz (6 Threads)

### Issue description

I want to put Metadata on a gd script. The Editor lets me do this:
![I have added test_meta:bool in this image.](https://github.com/godotengine/godot/assets/392218/26cdae12-aa86-4e86-9f4a-da50bb81dffe)

Saving the project, closing and re-loading it results in this:
![The metadata is gone.](https://github.com/godotengine/godot/assets/392218/3f7de89b-886c-49af-9c66-3b09e7592768)

### Steps to reproduce

Create a gd script. Double-click it. Go into the Inspector and add metadata. Save project. Close. Open the project and go back to that script in the inspector.

### Minimal reproduction project

None required.

**Comment by YuriSizov** [comment]
This is expected. GDScript is saved as a plain script file, it doesn't have a place to store metadata. Unlike resources saved as RES/TRES, which can save arbitrary data with them.

**Comment by donn-xx** [comment]
Well, it would be you, the dev, making the metadata, so not that surprising. However, I bow to your experience.

**Comment by dalexeev** [comment]
* See also #82333 and #82406.

**Comment by Johnrobmiller** [comment]
> This is expected. GDScript is saved as a plain script file, it doesn't have a place to store metadata. Unlike resources saved as RES/TRES, which can save arbitrary data with them.
> 
> We should probably disable the section on the inspector for scripts, and document this somewhere.

Be careful when saying things like "This is expected". For whom is it expected for? Is it expected for the developer, who implemented the solution and has the correct mental model of the system's behavior? Or, it is expected for the user, who was never provided with an interface with good enough usability for them to learn the correct mental model.

It is expected for the former, but not the latter.

There are a huge number of things in Godot that ought to be redesigned (by real UX designers, not programmers) such that it is expected from both perspectives.

**Comment by YuriSizov** [comment]
It is the expected behavior of the system given the limitations and considerations that is contains. Whether it is intuitive or not, whether it is convenient or not, whether it is communicated well or not are all different questions. When we say "This is expected" we are teaching the user that this is what you should expect to happen in this case. As learning the system as a part of building that mental model that you are speaking of.

In any case, I don't think sophism is helpful here.

**Comment by Calinou** [comment]
Note that this limitation applies to any resource type that is saved as a file type that isn't derived from `.res` or `.tres`.

This means that text-based `.gdshader` can't contain persistent metadata, but VisualShader can as it uses the `.tres` format.

**Comment by Zireael07** [comment]
1) It's worth changing the issue title so that people know it's not limited to gdscript
2) It's worth documenting unless you act fast and hide act metadata button before next release

**Comment by Johnrobmiller** [comment]
> It is the expected behavior of the system given the limitations and considerations that is contains. Whether it is intuitive or not, whether it is convenient or not, whether it is communicated well or not are all different questions. When we say "This is expected" we are teaching the user that this is what they should expect to happen in this case. As learning the system is a part of building that mental model that you are speaking of.
> 
> In any case, I don't think sophism is helpful here. Let's stick to the technical argument.

Fair enough. It may be pedantic, but I still make the point because engineers have a tendency to view things from the perspective of an engineer (as they should), but problems arise when they fail to view things from the perspective of a user, which is required to engineer the usability of a product.

Terms such as "expected behavior" only reinforce these bad habits, and so it's a pet peeve of mine when they are used.

If it's a big company and there are dedicated designers and dedicated engineers, then engineers don't need to think about usability. However, when the engineers are also the designers, they need to spent significantly more effort trying to forget about their engineering knowledge and view things purely from the users perspective so that they can design.

**Comment by Zehir** [comment]
I just found this issue looking why the metadata was not saved.
Since 4.5 there is now a .uid file. I guess we could store the metadata in that file.
Or just hide the add metadata button on text file like mentioned before.
I know it's not really a priority to fix that issue but I fell for it like other users.
